In [1]:
import matplotlib.pyplot as plt
import scipy.stats as st
import seaborn as sns
import pandas as pd
import numpy as np

# Aufgabe 3.1
Die monatlichen Aufwendungen X [CHF] für denWasserverbrauch eines 2-Personenhaushalts seien durch eine Zufallsvariable mit der folgenden Wahrscheinlichkeitsdichtefunktion gegeben:
$$\int_{0}^{60} cx(15-\frac{x}4 ) dx$$ 
### a) Welchen Wert muss c annehmen?

mit Wahrscheinlichkeitsdichtefunktion das Integral bestimmen $$c\int_{0}^{60} (15x-\frac{x^2}4 ) dx$$
und ausrechnen


In [8]:
from scipy.integrate import quad
f = lambda x: x*(15-x/4)
ans, _ = quad(f, 0, 60)

In [9]:
print(ans)

9000.0


Dies entspricht dem Integral ohne Konstante c<br>
=> $c * 9000 = 1$, da die Wahrscheinlichkeitsdichtefunktion immer 1 ergibt<br>

$c = \frac{1}{9000}$

### b) Geben Sie die Verteilungsfunktion F der Zufallsvariablen X an.
Das heisst integrieren (aufleiten)

Aus $$F(t) = \frac{1}{9000}\int_{0}^{60} (15x-\frac{x^2}4 ) dx$$ wird 
$$F(t) = \frac{1}{9000} * (\frac{15}{2}x^2 -\frac{1}{12}x^3)$$
$F(x)$ zu $F(t)$ umbennant, damit keine Verwirrung auftritt

### c) Welcher Wert der Aufwendungen wird nur mit 10% Wahrscheinlichkeit überschritten?

eine 10% Überschreitung = 90% tritt ein<br>
P(X > a) = 10% oder P(X <= a) = 90%

 $$0.9 = \frac{1}{9000} * (\frac{15}{2}x^2 -\frac{1}{12}x^3)$$
#### Solve muss = 0 gesetzt werden, da Python nicht im Stande ist die Gleichung sonst zu lösen (--> TI-nspire brauchen)
 $$0 = \frac{1}{9000} * (\frac{15}{2}x^2 -\frac{1}{12}x^3) - 0.9$$

In [28]:
from sympy.solvers import solve
from sympy import Symbol
x = Symbol('x')
solve((1/9000)*((15/2)*x**2-(1/12)*x**3)-0.9, x)

[-28.6249720273366 - 0.e-22*I,
 48.2519936604545 + 0.e-19*I,
 70.3729783668821 + 0.e-19*I]

Die Lösung ist 48.25, da weder -28 noch 70 im Bereich 0-60 sind.

### d) Wie hoch sind die erwarteten monatlichen Aufwendungen für den Wasserverbrauch eines 2-Personenhaushalts?

erwartet --> Erwartungswert

Erwartungswert: $E(X) = \int_{0}^{60} x f(x)dx$ und $f(x) = \frac{1}{9000}\int_{0}^{60} (15x-\frac{x^2}4 ) dx$ <br><br>--><br>
$E(X) = \int_{0}^{60} x * \frac{1}{9000} (15x-\frac{x^2}4 ) dx$ <br>--> <br>
$E(X) = \frac{1}{9000} \int_{0}^{60} x * (15x-\frac{x^2}4 ) dx$ <br>--><br>
$$E(X) = \frac{1}{9000} \int_{0}^{60} x^2 (15-\frac{x}4 ) dx$$


In [30]:
f = lambda x: (1/9000) * x**2 *(15-x/4)
ans, _ = quad(f, 0, 60)

In [31]:
print(ans)

30.000000000000007


Der Wert beträgt 30

---
# Aufgabe 3.2
In der Stadt Luzern gibt es bekanntlich viele Baustellen. Die Dauer X der Arbeiten bei einer Baustelle liege zwischen 0 und 20 Wochen. Die Dichte f (x) habe die folgende Form.
### a) Begründen Sie, warum c = 0.1 ist und schreiben Sie die Dichte f (x) explizit auf.

$0.1 * 20 = 2$ <br>
Die Grafik ist aber ein Dreieck --> <br>
$$ \frac{0.1*20}2 = 1$$

Dichte f(x) explizit:<br>
$$\frac{c * 20}2 = 1$$
--> $$c = \frac{1}{10}$$

$f(x)$ für [0 <= x <= 20] -->$\frac{1}{10}*(1-\frac{x}{20})$<br>
$f(x)$ für [x < 0 & x > 20] -->$0$

### b) Berechnen Sie die Wahrscheinlichkeit, dass die Bauzeit X weniger als a) 5, b) 10 Wochen beträgt.

Integriere f(x) um F(x) zu erhalten

$$\frac{1}{10}*(1-\frac{x}{20})$$
-->
$$f(x) = \int_0^n(\frac{1}{10}-\frac{x}{200})dx $$
-->
$$F(x) = \frac{x}{10}-\frac{x^2}{400}$$

**Wissenswert**: x <= 0 --> F(x) = 0 & x >= 20 --> F(x) = 1

5 Wochen

In [48]:
(5/10)-(5**2/400)

0.4375

10 Wochen

In [49]:
(10/10)-(10**2/400)

0.75

### c) Skizzieren Sie die kumulative Verteilungsfunktion.

siehe Skizze im Buch

### d) Berechnen Sie den Erwartungswert, den Median und die Standardabweichung der Dauer X.

#### Erwartungswert

$$E(X) = \int_{-\infty}^{\infty}xf(x)dx$$

$$E(X) = \int_{0}^{20} x (\frac{1}{10}-\frac{x}{200})dx$$

In [55]:
from scipy.integrate import quad
f = lambda x: x*((1/10)-(x/200))
ans3, _ = quad(f, 0, 20)
print(ans3)

6.666666666666668


#### Standardabweichung durch Varianz
$$VAR(X) = E(X^2) - E(X)^2$$

$$E(X^2) = \int_{-\infty}^{\infty}x^2f(x)dx$$
$$E(X^2) = \int_{0}^{20} x^2 (\frac{1}{10}-\frac{x}{200})dx$$

$E(X^2)$

In [64]:
from fractions import Fraction
f = lambda x: (x**2)*((1/10)-(x/200))
ans4, _ = quad(f, 0, 20)
print(ans4)

66.66666666666669


In [66]:
Fraction(ans4).limit_denominator()

Fraction(200, 3)

$E(X)^2$

In [59]:
ans32=ans3**2

In [67]:
Fraction(ans32).limit_denominator()

Fraction(400, 9)

$VAR$

In [68]:
ans5 = ans4-ans32

In [69]:
Fraction(ans5).limit_denominator()

Fraction(200, 9)

Standardabweichung $\varphi^2$ = Varianz<br>
$\varphi$ = $\sqrt{(Varianz)}$

In [74]:
np.sqrt(ans5)

4.714045207910318

#### Median

F(Median) = 0,5 --> Median liegt also zwingend zwischen 0 - 20

$$0.5 = \frac{x}{10}-\frac{x^2}{400}$$

In [76]:
solve((x/10)-(x**2/400)-0.5)

[5.85786437626905, 34.1421356237310]

da 34 nicht zwischen 0-20 liegt ist 5.85 der Median.

### e) $ K = 40000 · \sqrt{X}$ entspreche dem Betrag in Franken, den die Arbeiten bei einer Baustelle kosten. Wie gross ist die Wahrscheinlichkeit, dass die Arbeiten bei einer Baustelle höchstens 120’000.- Fr. kosten?

$$P(K\leq120'000) = P(40000 * \sqrt{X} \leq 120'000)$$
<center>auflösen</center>
$$P(X \leq 9)$$
$$F(9)$$

einsetzen in Formel F(x) von oben

$$\frac{9}{10}-\frac{9^2}{400}$$

In [79]:
(9/10)-(9**2/400)

0.6975

Die W'keit beträgt 69.75%

Die vorgeschlagene Verteilung ist nur ein Modell. Man könnte die Dauer der Bauarbeiten zum Beispiel auch als exponentialverteilt annehmen.
### f) Für welchen Parameter λ hat die Exponentialverteilung denselben Erwartungswert wie die bisherige angenommene Verteilung?

$$E(X) = \frac{1}{\lambda}$$

$$\int_{0}^{20} x (\frac{1}{10}-\frac{x}{200})dx = \frac{1}{\lambda}$$

In [84]:
f = lambda x: x*((1/10)-(x/200))
erW, _ = quad(f, 0, 20)
Fraction(erW).limit_denominator()

Fraction(20, 3)

$$\frac{20}3 = \frac{1}{\lambda}$$

In [90]:
solve((20/3)-(1/x),x)

[0.150000000000000]

In [91]:
Fraction(0.15).limit_denominator()

Fraction(3, 20)

Der Erwartungswert ist 0.15 oder $\frac{3}{20}$

### g) Berechnen Sie mit der gefundenen Exponentialverteilung nochmals Teilaufgabe e).

Änderung: statt F(9) gibt es jetzt z.B.: G(9) welches wir in die exponentionelle kumulative Verteilungsfunktion einsetzen
$$1 - \exp({-\frac{3}{20}*9})$$

In [96]:
from sympy.functions import exp
1-exp(-(3/20)*9)

0.740759739354108

mit einer Exponentialverteilung sind die Chancen höher dass die Arbeiten unter 120'000 bleiben.

--- 
# Aufgabe 3.3
Eine Maschine hat eine exponentialverteilte Lebensdauer T1 (in Stunden) mit Parameter λ = 0.04.
### a) Berechnen Sie den Median und den Erwartungswert. Mit welcher Wahrscheinlichkeit überlebt das System seine Lebenserwartung?

$$E(X) = \frac{1}{\lambda} = \frac{1}{0.04} = 25$$

$$Median = \frac{\ln{(2)}}{\lambda}$$

In [98]:
np.log(2)/0.04

17.328679513998633

$$Lebenserwartung = P(X \gt E(X)) = P(X \gt 25)$$
$$= P(X\gt25)$$
$$= 1 - P(X\gt25)$$
<center>da die W'keit grösser als der E'Wert sein soll, setzen wir gleich diesen ein (Siehe Skizze im Buch)</center>
$$= 1 - F(25) --> 1 - F(\frac{1}{\lambda})$$

$$\frac{1}{\lambda} = 1 - \exp({-\lambda*\frac{1}{\lambda}})$$
$$ = 1 - \exp({\frac{-\lambda}{1}*\frac{1}{\lambda}})$$
$$ = 1 - \exp({\frac{-\lambda}{\lambda}})$$
$$ = 1 - \exp({-1})$$
<br>
Das einsetzen in die obere Gleichung
<br>
$$1 - (1 - \exp({-1})$$
$$= 1 - 1 + \exp({-1})$$
$$=\exp({-1})$$


In [106]:
np.exp(-1)

0.36787944117144233

Mit ca. 36% überlebt das System seine Erwartung

Variante 2: mit f(x) statt F(x)

In [107]:
f = lambda x: 0.04*exp(-0.04*x)
ld, _ = quad(f,0,25)
ld = 1 - ld
print(ld)

0.3678794411714422


### b) Mit welcher Wahrscheinlichkeit liegt die Lebensdauer des Systems im Bereich E[T1] ± σT1?

<center>$$\mu = \sigma = \frac{1}{\lambda}$$, bei Exponentialverteilung</center>

Die Dauer die wir berechnen wollen liegt nun zwischen $E(X)-\sigma$ bis $E(X)+\sigma$ = $$\mu-\sigma \leq X \leq \mu+\sigma$$

$$P(0 \leq X \leq \frac{2}{\lambda})$$
$$= F(\frac{2}{\lambda}) - F(0)$$
$$= [1-\exp({-2})] - 0$$

In [109]:
(1-np.exp(-2))-0

0.8646647167633873

W'keit: 86%

### c) Nun betrachten wir eine zweite Maschine mit der Lebensdauer T2 (in Stunden) mit der folgenden Dichte (siehe Aufgabenstellung). Wie gross muss c2 sein? Berechnen Sie den Erwartungswert und die Standardabweichung der Zufallsvariablen T2.

Siehe Notizbuch
$$c2 = \frac{1}{1000}$$

$$E(X) = \frac{1}{\frac{1}{1000}}$$

In [111]:
1/(1/1000)

1000.0

$$\mu = \sigma = 1000$$

### d) Wir betrachten nun das System bestehend aus den beiden Maschinen mit Lebensdauern T1 und T2, welche unabhängig voneinander funktionieren. Berechnen Sie die Wahrscheinlichkeit, dass nach 200 Stunden beide Maschinen noch funktionieren.

#### T1 & T2
$$[1 - FT1(200)] * [1 - FT2(200)]$$
kum. Verteilung: $1 - \exp(- \lambda x)$ = F(x)

$\lambda$ = c --> E(X) = $\frac{1}{\lambda}$ gem. Wahrscheinlichkeitsdichte

$$FT1(200) = 1 - \exp({-0.04*200})$$
$$T1 = 1 - [1 - \exp({-0.04*200})]$$
$$T1 = \exp({-0.04*200})$$

$$FT2(200) = 1 - \exp({-(1/1000)*200})$$
$$T2 = 1 - [1 - \exp({-(1/1000)*200})]$$
$$T2 = \exp({-(1/1000)*200})$$

In [121]:
T1 = exp(-0.04*200)

In [122]:
T2 = exp(-(1/1000)*200)

In [123]:
T1 * T2

0.000274653569972142

0.027 %

---
# Aufgabe 3.4
Aufgrund langjähriger Untersuchungen ist bekannt, dass der Bleigehalt X in einer Bodenprobe annähernd normalverteilt ist.  Ausserdem weiss man, dass der Erwartungswert 32 ppb beträgt und dass die Standardabweichung 6 ppb beträgt. 
### a) Machen Sie eine Skizze der Dichte von X, und zeichnen Sie die Wahrscheinlichkeit, dass eine Bodenprobe zwischen 26 und 38 ppb Blei enthält, in die Skizze ein.

Siehe Skizze im Notizbuch

### b) Wie gross ist die Wahrscheinlichkeit, dass eine Bodenprobe höchstens 40 ppb Schwermetall enthält? 
Hinweis: Benützen Sie die Python-Funktion norm.cdf()<br> <br>
Überprüfen Sie das Resultat, indem Sie die standardisierte Zufallsvariable Z einführen und mit der Standardnormalverteilung die Wahrscheinlichkeit berechnen.

$$ P(X\leq40)$$

In [126]:
from scipy.stats import uniform, expon, norm
norm.cdf(x=40, loc=32, scale=6)

0.9087887802741321

Überprüfen des Resultats mit umwandeln in standardisierte Zufallsvariable kommt erst in Kapitel 3.4 vor.<br>
**Formel zum Umrechnen:**<br>
<center>$Z = \frac{X-\mu}{\sigma}$ ~ $N(0,1)$</center>


<center>$Z = \frac{X - 32}{6}$ ~ $N(0,1)$</center><br>
<center>$Z = \frac{40 - 32}{6}$ ~ $N(0,1)$</center><br>
<center>$Z = \frac{8}{6}$ ~ $N(0,1)$</center><br>
<center>$Z = \frac{4}{3}$ ~ $N(0,1)$</center>

In [127]:
norm.cdf(x=(4/3))

0.9087887802741321

### c) Wie gross ist die Wahrscheinlichkeit, dass eine Bodenprobe höchstens 27 ppb Schwermetall enthält?

In [128]:
norm.cdf(x=27, loc=32, scale=6)

0.20232838096364308

### d) Welcher Bleigehalt wird mit einer Wahrscheinlichkeit von 97.5% unterschritten?
Das heisst, bestimmen Sie dasjenige c, so dass die Wahrscheinlichkeit, dass der Bleigehalt kleiner oder gleich c ist, genau 97.5% beträgt. Hinweis: Benützen Sie die Python -Funktion norm.ppf()

In [130]:
norm.ppf(q=0.975, loc=32, scale=6)

43.759783907240326

### e) Welcher Bleigehalt wird mit einer Wahrscheinlichkeit von 10% unterschritten?

In [131]:
norm.ppf(q=0.1, loc=32, scale=6)

24.310690606732397

### f) Wie gross ist die Wahrscheinlichkeit, die in Aufgabe a) eingezeichnet wurde?

In [132]:
norm.cdf(x=38, loc=32, scale=6) - norm.cdf(x=26, loc=32, scale=6)

0.6826894921370859

---
# Aufgabe 3.5
Wir gehen davon aus, dass bei der Detektion eines digitalen Signals das Hintergrundrauschen einer Normalverteilung folgt, und zwar mit einem Mittelwert von 0 Volt und einer Standardabweichung von 0.45 Volt. Das System geht davon aus, dass eine digitale 1 übertragen worden ist, wenn die Spannung 0.9 Volt überstiegen worden ist.
### a) Wie gross ist die Wahrscheinlichkeit, dass eine 1 detektiert worden ist, wenn in Wahrheit keine gesendet wurde? Wie interpretieren Sie diese Wahrscheinlichkeit?

<center>2x $\sigma$ = 0.9 --> Signal übertragen</center><br>
$$P(\mu - 2\sigma \leq X \leq \mu + 2\sigma) \approx 0.95$$<br>
<center>Die Wahrscheinlichkeit sollte 0.275% betragen, d.h. die Wahrscheinlichkeit ist nichtig.

### b) Bestimmen Sie die symmetrischen Grenzen um 0 Volt, die 99% des Hintergrundrauschens einschliessen.

99% um den Mittelwert --> 99:2 = 49.5% auf beide Seiten<br>
Quantil_1 = 0.5<br>
Quantil_2 = 99.5<br>

In [135]:
norm.ppf(q=[0.005,0.995], loc=0, scale=0.45)

array([-1.15912319,  1.15912319])

### c) Nehmen Sie an, eine digitale 1 wird als eine Verschiebung des Mittelwertes um 1.8 Volt im Vergleich zum Rauschsignal dargestellt. Die Standardabweichung des Signals ist ebenfalls 0.45 Volt. Wie gross ist die Wahrscheinlichkeit, dass eine digitale 1 nicht detektiert wird?

Eine digitale 1 wird bei 0.9 Volt detektiert (siehe oben) --> $P(X \lt 0.9)$

In [137]:
norm.cdf(x=0.9, loc=1.8, scale=0.45)

0.022750131948179195

---
# Aufgabe 3.6
Der Schaftdurchmesser von optischen Speicherplatten ist normalverteilt mit einem Mittelwert von 0.2508 Millimeter und Standardabweichung 0.0005 Millimeter. Die technischen Angaben des Schafts sind 0.2500± 0.0015 Millimeter.
### a) Welcher Anteil der Schaftdurchmesser entspricht den technischen Angaben?

technische Angaben unten = 0.2500 - 0.0015 mm<br>
technische Angaben oben = 0.2500 + 0.015mm<br>
0.2485 - 0.2515 Range

In [138]:
norm.cdf(x=0.2515, loc=0.2508, scale=0.0005) - norm.cdf(x=0.2485, loc=0.2508, scale=0.0005)

0.91924122831152

### b) Die meisten Schaftdurchmesser sind zu gross, da der Mittelwert der Herstellung sehr nahe an der oberen Grenze der technischen Angabe liegt. Wenn der Herstellungsprozess derart geändert wird, dass der Mittelwert der Schaftdurchmesser nun beim Zielwert von 0.2500 liegt,welcher Anteil entspricht dann den technischen Angaben?

In [139]:
norm.cdf(x=0.2515, loc=0.2500, scale=0.0005) - norm.cdf(x=0.2485, loc=0.2500, scale=0.0005)

0.9973002039367398

---
# Aufgabe 3.7
Monte Carlo Algorithmen sind randomisierte Algorithmen und stellen ein gutes Werkzeug für Simulationen von stochastischen Prozessen dar. Auch die Zahl π lässt sich mit Hilfe von Monte Carlo Simulationen bestimmen. Im Folgenden möchten wir ein Computerprogramm erstellen, mit welchem man die Zahl π aufgrund von Monte Carlo Methoden simulieren kann. Man generiert hierzu zufällige Punkte<br><br>
$$P ∈ {(x, y)|x ∈ [−1, 1] und y ∈ [−1, 1]}$$<br>
und überprüft, ob diese innerhalb des Einheitskreises mit Kreismittelpunkt MK = (0, 0) und Radius r = 1 liegen. Die sich ergebende Wahrscheinlichkeitsverteilung $P[(x, y) ∈ Kreis]$ stellt die Fläche eines Viertels des Einheitskreises dar. π kann nun mit folgender Formel berechnet werden<br><br>
$$\frac{Anzahl  Treffer  in  Kreisfläche}{Anzahl  generierter  Punkte  Quadrat} = \frac{Kreisfläche}{Quadratfläche} = \frac{r^2*\pi}{(2*r)^2} = \frac{\pi}{4} = P[(x,y) \epsilon Kreis].$$<br>
Bestimmen Sie mit Hilfe dieser Überlegung die Zahl $\pi$.


#### Was ich verstanden habe:
x ist Element von -1 - 1<br>
y ist Element von -1 - 1<br>
Einheitskreismittelpunkt-Koordinaten: (0,0)<br>
Radius Einheitskreis: r=1<br>

In [140]:
# Generieren von 100 gleichmaessig verteilten
# Zufallszahlen im Intervall [-1,1]
from scipy.stats import uniform
import numpy as np
x = uniform.rvs(size=100, loc=-1, scale=2)
# Bestimmen der Anzahl Zahlen kleiner als eins; Beispiel:
# Anzahl von 100 zufaellig im Intervall [0,10] generierten
# Zahlen, die kleiner als 1 sind:
np.sum(x<1)

100

# Wieso Erwartungswert (loc) von -1? 
# Woher weiss uniform was x und y Achse ist?

# Wieso X*X + Y*Y <-- Sinn?